### PR TITLE
[Feature](Nereids) Support hll and count for materialized index.

### PR DESCRIPTION
**Pull request by wangshuo128**
# Proposed changes

This pr add rewrite for HLL and count for materialized index.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.64 seconds
 load time: 637 seconds
 storage size: 17123266935 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221222081303_clickbench_pr_67141.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java [hunk 1]
```diff
@@ -328,6 +330,13 @@ private SelectResult select(
                         .filter(index -> !candidatesWithoutRewriting.contains(index))
                         .map(index -> rewriteAgg(index, scan, requiredScanOutput, predicates, aggregateFunctions,
                                 groupingExprs))
+                        .filter(aggRewriteResult -> checkPreAggStatus(scan, aggRewriteResult.index.getId(),
+                                predicates,
+                                // check pre-agg status of aggregate function that couldn't rewrite.
+                                // ImmutableList.copyOf(Sets.difference(ImmutableSet.copyOf(aggregateFunctions),
+                                //         aggRewriteResult.exprRewriteMap.aggFuncMap.keySet())),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java:337]: useless code

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Ndv.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            FunctionSignature.ret(BigIntType.INSTANCE).args(DateType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(DateTimeType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(DateV2Type.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(DateTimeV2Type.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(TimeType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(TimeV2Type.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(JsonType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(HllType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(BitmapType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(QuantileStateType.INSTANCE)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Ndv.java:84]: legacy planner also support NULL_TYPE and CharType

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java [hunk 3]
```diff
@@ -711,34 +746,62 @@ private static Expression rewrite(Expression expr, RewriteContext context) {
 
         /**
          * count(distinct col) -> bitmap_union_count(mv_bitmap_union_col)
+         * count(col) -> sum(mv_count_col)
          */
         @Override
         public Expression visitCount(Count count, RewriteContext context) {
             if (count.isDistinct() && count.arity() == 1) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java:753]: should the AggFuncRewriter process children first in each visit function? such as
```
public Expression visitCount(Count count, RewriteContext context) {
    List<Expression> children = count.children().stream().map(e -> this.rewrite(e, context)).collect(Collectors.toList());
    ...
}
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java (hunk 1 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java]: thanks, fixed.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Ndv.java (hunk 2 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Ndv.java]: fixed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java (hunk 3 above)
Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/mv/SelectMaterializedIndexWithAggregate.java:723]: This rewriter is a bit different. 
It only handles aggregate function on a slot, e.g., rewriting `count(col)` to `sum(mv_count_col)`. The child of the aggregate function is checked by `ExpressionUtils.extractSlotOrCastOnSlot(count.child(0))` to ensure it's a slot.
 So we don't need to rewrite children first.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Ndv.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            FunctionSignature.ret(BigIntType.INSTANCE).args(DateV2Type.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(DateTimeV2Type.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(TimeType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(TimeV2Type.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(JsonType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(HllType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(BitmapType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(QuantileStateType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(NullType.INSTANCE),
+            FunctionSignature.ret(BigIntType.INSTANCE).args(CharType.INSTANCE)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Ndv.java:88]: it is a bit hard to use here.
`CharType.INSTANCE` is return `CharacterType` actually.
We need use `CharType.SYSTEM_DEFAULT` just like `VarcharType`

Comment (wangshuo128) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/Ndv.java]: fixed.

**Review by morrySnow — APPROVED** [review]

**Pull request merged**